File: PluralKit.Core/Models/SystemConfig.cs
using Newtonsoft.Json.Linq;

using NodaTime;

namespace PluralKit.Core;

public class SystemConfig
{
    public SystemId Id { get; }
    public string UiTz { get; set; }
    public bool PingsEnabled { get; }
    public int? LatchTimeout { get; }
    public bool MemberDefaultPrivate { get; }
    public bool GroupDefaultPrivate { get; }
    public bool ShowPrivateInfo { get; }
    public int? MemberLimitOverride { get; }
    public int? GroupLimitOverride { get; }
    public ICollection<string> DescriptionTemplates { get; }

    public DateTimeZone Zone => DateTimeZoneProviders.Tzdb.GetZoneOrNull(UiTz);

    public bool CaseSensitiveProxyTags { get; }
    public bool ProxyErrorMessageEnabled { get; }
    public bool HidDisplaySplit { get; }
    public bool HidDisplayCaps { get; }
    public bool CardShowColorHex { get; }
    public HidPadFormat HidListPadding { get; }
    public ProxySwitchAction ProxySwitch { get; }
    public string NameFormat { get; }

    public bool PremiumLifetime { get; }
    public Instant? PremiumUntil { get; }

    public enum HidPadFormat
    {
        None = 0,
        Left = 1,
        Right = 2,
    }
    public enum ProxySwitchAction
    {
        Off = 0,
        New = 1,
        Add = 2,
    }
}

public static class SystemConfigExt
{
    public static JObject ToJson(this SystemConfig cfg)
    {
        var o = new JObject();

        o.Add("timezone", cfg.UiTz);
        o.Add("pings_enabled", cfg.PingsEnabled);
        o.Add("latch_timeout", cfg.LatchTimeout);
        o.Add("member_default_private", cfg.MemberDefaultPrivate);
        o.Add("group_default_private", cfg.GroupDefaultPrivate);
        o.Add("show_private_info", cfg.ShowPrivateInfo);
        o.Add("member_limit", cfg.MemberLimitOverride ?? Limits.MaxMemberCount);
        o.Add("group_limit", cfg.GroupLimitOverride ?? Limits.MaxGroupCount);
        o.Add("case_sensitive_proxy_tags", cfg.CaseSensitiveProxyTags);
        o.Add("proxy_error_message_enabled", cfg.ProxyErrorMessageEnabled);
        o.Add("hid_display_split", cfg.HidDisplaySplit);
        o.Add("hid_display_caps", cfg.HidDisplayCaps);
        o.Add("hid_list_padding", cfg.HidListPadding.ToUserString());
        o.Add("card_show_color_hex", cfg.CardShowColorHex);
        o.Add("proxy_switch", cfg.ProxySwitch.ToUserString());
        o.Add("name_format", cfg.NameFormat);

        o.Add("description_templates", JArray.FromObject(cfg.DescriptionTemplates));

        return o;
    }

    public static string ToUserString(this SystemConfig.HidPadFormat val)
    {
        if (val == SystemConfig.HidPadFormat.None) return "off";
        return val.ToString().ToLower();
    }

    public static string ToUserString(this SystemConfig.ProxySwitchAction val) => val.ToString().ToLower();

}

File: crates/models/src/system_config.rs
use pk_macros::pk_model;

use sqlx::{postgres::PgTypeInfo, Database, Decode, Postgres, Type};
use std::error::Error;

use crate::{SystemId, _util::fake_enum_impls};

pub const DEFAULT_MEMBER_LIMIT: i32 = 1000;
pub const DEFAULT_GROUP_LIMIT: i32 = 250;

#[derive(serde::Serialize, Debug, Clone)]
#[serde(rename_all = "snake_case")]
pub enum HidPadFormat {
    #[serde(rename = "off")]
    None,
    Left,
    Right,
}
fake_enum_impls!(HidPadFormat);

impl From<i32> for HidPadFormat {
    fn from(value: i32) -> Self {
        match value {
            0 => HidPadFormat::None,
            1 => HidPadFormat::Left,
            2 => HidPadFormat::Right,
            _ => unreachable!(),
        }
    }
}

#[derive(serde::Serialize, Debug, Clone)]
#[serde(rename_all = "snake_case")]
pub enum ProxySwitchAction {
    Off,
    New,
    Add,
}
fake_enum_impls!(ProxySwitchAction);

impl From<i32> for ProxySwitchAction {
    fn from(value: i32) -> Self {
        match value {
            0 => ProxySwitchAction::Off,
            1 => ProxySwitchAction::New,
            2 => ProxySwitchAction::Add,
            _ => unreachable!(),
        }
    }
}

#[pk_model]
struct SystemConfig {
    system: SystemId,
    #[json = "timezone"]
    ui_tz: String,
    #[json = "pings_enabled"]
    pings_enabled: bool,
    #[json = "latch_timeout"]
    latch_timeout: Option<i32>,
    #[json = "member_default_private"]
    member_default_private: bool,
    #[json = "group_default_private"]
    group_default_private: bool,
    #[json = "show_private_info"]
    show_private_info: bool,
    #[json = "member_limit"]
    #[default = DEFAULT_MEMBER_LIMIT]
    member_limit_override: Option<i32>,
    #[json = "group_limit"]
    #[default = DEFAULT_GROUP_LIMIT]
    group_limit_override: Option<i32>,
    #[json = "case_sensitive_proxy_tags"]
    case_sensitive_proxy_tags: bool,
    #[json = "proxy_error_message_enabled"]
    proxy_error_message_enabled: bool,
    #[json = "hid_display_split"]
    hid_display_split: bool,
    #[json = "hid_display_caps"]
    hid_display_caps: bool,
    #[json = "hid_list_padding"]
    hid_list_padding: HidPadFormat,
    #[json = "proxy_switch"]
    proxy_switch: ProxySwitchAction,
    #[json = "name_format"]
    #[default = "{name} {tag}".to_string()]
    name_format: Option<String>,
    #[json = "description_templates"]
    description_templates: Vec<String>,
}
